File: shuffle/shuffle.py
fin = open('shuffle.in')
data = fin.readlines()
fin.close()
N = int(data[0][:-1])
a = [int(i) for i in data[1][:-1].split(' ')]

def followPath(index):
	visited = [index]
	while a[visited[-1]]-1 not in visited:
		visited.append(a[visited[-1]]-1)
	return (a[visited[-1]]-1 == index)

fout = open('shuffle.out', 'w')
fout.write(str(sum([followPath(i) for i in range(N)])))
fout.close()

File: shuffle/shuffle.cpp
#include <iostream>
#include <fstream>
#include <algorithm>
#include <iterator>
using namespace std;

int main(void) {
	ifstream fin;
	fin.open("shuffle.in");
	int N;
	fin>>N;
	int a[N];
	for (int i = 0; i<N; i++) {
		fin>>a[N];
	}
	fin.close();
	
	int tmp[N] = {-1};
	int path = 0;
	int ans = 0;
	for (int i = 0; i<N; i++) {
		path = i;
		int counter = 0;
		tmp[N] = {-1};
		while (find(tmp, tmp+N, a[path]-1) == tmp+N) {
			path = a[path]-1;
			tmp[counter] = path;
			counter++;
		}
		if (a[path]-1 == i) {
			ans++;
		}
	}

	ofstream fout;
	fout.open("shuffle.out");
	fout<<ans<<endl;
	fout.close();
	return 0;
}
